<a href="https://colab.research.google.com/github/Titas23/IS-170-Machine-Learning/blob/Association-Rules/HW_5_AndresP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#set up the libraries
# we use apriori for SUPPORT and association_rules for CONFIDENCE, LIFT and CONVICTION
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

In [ ]:
#Loading the dataset
df=pd.read_csv('dataset1.csv')
df.head()

Milk, Onion, Nutmeg, Kidney Beans, Eggs, Yogurt
0    Dill, Onion, Nutmeg, Kidney Beans, Eggs, Yogurt
1                    Milk, Apple, Kidney Beans, Eggs
2          Milk, Unicorn, Corn, Kidney Beans, Yogurt
3  Corn, Onion, Onion, Kidney Beans, Ice cream, Eggs
4            Blueberries, Banana, Flour, Sugar, Eggs

In [ ]:
df.isnull().sum()
df.loc[(df!=0).any(axis=1)]

Milk, Onion, Nutmeg, Kidney Beans, Eggs, Yogurt
0    Dill, Onion, Nutmeg, Kidney Beans, Eggs, Yogurt
1                    Milk, Apple, Kidney Beans, Eggs
2          Milk, Unicorn, Corn, Kidney Beans, Yogurt
3  Corn, Onion, Onion, Kidney Beans, Ice cream, Eggs
4            Blueberries, Banana, Flour, Sugar, Eggs
5                 Oranges, Bread, Flour, Eggs, Sugar
6             Diapers, wipes, Baby Food, Apple Juice
7             Diapers, wipes, Apple Juice, Baby Food
8             wipes, Diapers, Apple Juice, Baby Food
9                Bread, Oranges,  Flour, Eggs, Sugar

In [ ]:
#Since it's missing a column header, so we create new column header name
df=pd.read_csv('dataset1.csv', names=['Groceries'], sep=',')
df.head(10)


Groceries
0    Milk, Onion, Nutmeg, Kidney Beans, Eggs, Yogurt
1    Dill, Onion, Nutmeg, Kidney Beans, Eggs, Yogurt
2                    Milk, Apple, Kidney Beans, Eggs
3          Milk, Unicorn, Corn, Kidney Beans, Yogurt
4  Corn, Onion, Onion, Kidney Beans, Ice cream, Eggs
5            Blueberries, Banana, Flour, Sugar, Eggs
6                 Oranges, Bread, Flour, Eggs, Sugar
7             Diapers, wipes, Baby Food, Apple Juice
8             Diapers, wipes, Apple Juice, Baby Food
9             wipes, Diapers, Apple Juice, Baby Food

In [ ]:
#data size
df.shape

(11, 1)

In [ ]:
#we are turning the pandas dataframe into a list of lists
data=list(df["Groceries"].apply(lambda x:x.split(",")))
data

[['Milk', ' Onion', ' Nutmeg', ' Kidney Beans', ' Eggs', ' Yogurt'],
 ['Dill', ' Onion', ' Nutmeg', ' Kidney Beans', ' Eggs', ' Yogurt'],
 ['Milk', ' Apple', ' Kidney Beans', ' Eggs'],
 ['Milk', ' Unicorn', ' Corn', ' Kidney Beans', ' Yogurt'],
 ['Corn', ' Onion', ' Onion', ' Kidney Beans', ' Ice cream', ' Eggs'],
 ['Blueberries', ' Banana', ' Flour', ' Sugar', ' Eggs'],
 ['Oranges', ' Bread', ' Flour', ' Eggs', ' Sugar'],
 ['Diapers', ' wipes', ' Baby Food', ' Apple Juice'],
 ['Diapers', ' wipes', ' Apple Juice', ' Baby Food'],
 ['wipes', ' Diapers', ' Apple Juice', ' Baby Food'],
 ['Bread', ' Oranges', '  Flour', ' Eggs', ' Sugar']]

In [ ]:
#Let's transform the list with one-hot encoding
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)
df=df.replace(True,1)
df=df.replace(False,0)
df

Flour   Apple   Apple Juice   Baby Food   Banana   Bread   Corn  \
0         0       0             0           0        0       0      0   
1         0       0             0           0        0       0      0   
2         0       1             0           0        0       0      0   
3         0       0             0           0        0       0      1   
4         0       0             0           0        0       0      0   
5         0       0             0           0        1       0      0   
6         0       0             0           0        0       1      0   
7         0       0             1           1        0       0      0   
8         0       0             1           1        0       0      0   
9         0       0             1           1        0       0      0   
10        1       0             0           0        0       0      0   

     Diapers   Eggs   Flour  ...   Yogurt   wipes  Blueberries  Bread  Corn  \
0          0      1       0  ...        1       0            0      0     0   
1          0      1       0  ...        1       0            0      0     0   
2          0      1       0  ...        0       0            0      0     0   
3          0      0       0  ...        1       0            0      0     0   
4          0      1       0  ...        0       0            0      0     1   
5          0      1       1  ...        0       0            1      0     0   
6          0      1       1  ...        0       0            0      0     0   
7          0      0       0  ...        0       1            0      0     0   
8          0      0       0  ...        0       1            0      0     0   
9          1      0       0  ...        0       0            0      0     0   
10         0      1       0  ...        0       0            0      1     0   

    Diapers  Dill  Milk  Oranges  wipes  
0         0     0     1        0      0  
1         0     1     0        0      0  
2         0     0     1        0      0  
3         0     0     1        0      0  
4         0     0     0        0      0  
5         0     0     0        0      0  
6         0     0     0        1      0  
7         1     0     0        0      0  
8         1     0     0        0      0  
9         0     0     0        0      1  
10        0     0     0        0      0  

[11 rows x 27 columns]

In [ ]:
#Set a threshold value for the support value and calculate the support value
df=apriori(df,min_support=0.1, use_colnames=True)
df

support                                          itemsets
0   0.272727                                    ( Apple Juice)
1   0.272727                                      ( Baby Food)
2   0.636364                                           ( Eggs)
3   0.181818                                          ( Flour)
4   0.454545                                   ( Kidney Beans)
5   0.181818                                         ( Nutmeg)
6   0.272727                                          ( Onion)
7   0.272727                                          ( Sugar)
8   0.272727                                         ( Yogurt)
9   0.181818                                          ( wipes)
10  0.181818                                         (Diapers)
11  0.272727                                            (Milk)
12  0.272727                        ( Apple Juice,  Baby Food)
13  0.181818                            ( Apple Juice,  wipes)
14  0.181818                           ( Apple Juice, Diapers)
15  0.181818                              ( Baby Food,  wipes)
16  0.181818                             (Diapers,  Baby Food)
17  0.181818                                   ( Flour,  Eggs)
18  0.363636                            ( Kidney Beans,  Eggs)
19  0.181818                                  ( Nutmeg,  Eggs)
20  0.272727                                   ( Onion,  Eggs)
21  0.272727                                   ( Eggs,  Sugar)
22  0.181818                                  ( Yogurt,  Eggs)
23  0.181818                                     ( Eggs, Milk)
24  0.181818                                  ( Flour,  Sugar)
25  0.181818                          ( Kidney Beans,  Nutmeg)
26  0.272727                           ( Kidney Beans,  Onion)
27  0.272727                          ( Yogurt,  Kidney Beans)
28  0.272727                             ( Kidney Beans, Milk)
29  0.181818                                 ( Nutmeg,  Onion)
30  0.181818                                ( Yogurt,  Nutmeg)
31  0.181818                                 ( Yogurt,  Onion)
32  0.181818                                   ( Yogurt, Milk)
33  0.181818                                 (Diapers,  wipes)
34  0.181818                ( Apple Juice,  Baby Food,  wipes)
35  0.181818               ( Apple Juice,  Baby Food, Diapers)
36  0.181818                   ( Apple Juice,  wipes, Diapers)
37  0.181818                     (Diapers,  Baby Food,  wipes)
38  0.181818                           ( Flour,  Eggs,  Sugar)
39  0.181818                   ( Kidney Beans,  Eggs,  Nutmeg)
40  0.272727                    ( Kidney Beans,  Eggs,  Onion)
41  0.181818                   ( Yogurt,  Eggs,  Kidney Beans)
42  0.181818                      ( Kidney Beans,  Eggs, Milk)
43  0.181818                          ( Nutmeg,  Eggs,  Onion)
44  0.181818                         ( Yogurt,  Eggs,  Nutmeg)
45  0.181818                          ( Yogurt,  Eggs,  Onion)
46  0.181818                  ( Kidney Beans,  Onion,  Nutmeg)
47  0.181818                 ( Yogurt,  Nutmeg,  Kidney Beans)
48  0.181818                  ( Yogurt,  Onion,  Kidney Beans)
49  0.181818                    ( Yogurt, Milk,  Kidney Beans)
50  0.181818                        ( Yogurt,  Onion,  Nutmeg)
51  0.181818       ( Apple Juice,  Baby Food,  wipes, Diapers)
52  0.181818           ( Kidney Beans,  Eggs,  Onion,  Nutmeg)
53  0.181818          ( Yogurt,  Eggs,  Nutmeg,  Kidney Beans)
54  0.181818           ( Yogurt,  Eggs,  Onion,  Kidney Beans)
55  0.181818                 ( Yogurt,  Eggs,  Onion,  Nutmeg)
56  0.181818         ( Yogurt,  Onion,  Nutmeg,  Kidney Beans)
57  0.181818  ( Yogurt,  Nutmeg,  Eggs,  Kidney Beans,  Onion)

Any item or combinations less than 1.0 are not shown.

In [ ]:
#Here is the CONFIDENCE (Try different min_threshold value to explore this more)
df_ar=association_rules(df,metric="confidence", min_threshold=0.3)
df_ar

antecedents                               consequents  \
0             ( Apple Juice)                              ( Baby Food)   
1               ( Baby Food)                            ( Apple Juice)   
2             ( Apple Juice)                                  ( wipes)   
3                   ( wipes)                            ( Apple Juice)   
4             ( Apple Juice)                                 (Diapers)   
..                       ...                                       ...   
239  ( Kidney Beans,  Onion)                 ( Yogurt,  Eggs,  Nutmeg)   
240                ( Yogurt)   ( Nutmeg,  Eggs,  Onion,  Kidney Beans)   
241                ( Nutmeg)   ( Yogurt,  Eggs,  Onion,  Kidney Beans)   
242          ( Kidney Beans)         ( Yogurt,  Eggs,  Onion,  Nutmeg)   
243                 ( Onion)  ( Yogurt,  Eggs,  Nutmeg,  Kidney Beans)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.272727            0.272727  0.272727    1.000000  3.666667   
1              0.272727            0.272727  0.272727    1.000000  3.666667   
2              0.272727            0.181818  0.181818    0.666667  3.666667   
3              0.181818            0.272727  0.181818    1.000000  3.666667   
4              0.272727            0.181818  0.181818    0.666667  3.666667   
..                  ...                 ...       ...         ...       ...   
239            0.272727            0.181818  0.181818    0.666667  3.666667   
240            0.272727            0.181818  0.181818    0.666667  3.666667   
241            0.181818            0.181818  0.181818    1.000000  5.500000   
242            0.454545            0.181818  0.181818    0.400000  2.200000   
243            0.272727            0.181818  0.181818    0.666667  3.666667   

     leverage  conviction  
0    0.198347         inf  
1    0.198347         inf  
2    0.132231    2.454545  
3    0.132231         inf  
4    0.132231    2.454545  
..        ...         ...  
239  0.132231    2.454545  
240  0.132231    2.454545  
241  0.148760         inf  
242  0.099174    1.363636  
243  0.132231    2.454545  

[244 rows x 9 columns]

The first two rows (Apple Juice -> Baby Food) shows a confidence of 1.0 which means the two items are bought together 100%. The third row (Eggs -> Kidney beans) shows a confidence of .57 which means these two items are bought together only 57% of the time.

In [ ]:
#Here is LIFT (try different min_threshold value to explore more) 
df_ar=association_rules(df,metric="lift", min_threshold=0.8)
df_ar

antecedents              consequents  antecedent support  \
0            ( Apple Juice)             ( Baby Food)            0.272727   
1              ( Baby Food)           ( Apple Juice)            0.272727   
2           ( Kidney Beans)                  ( Eggs)            0.454545   
3                   ( Eggs)          ( Kidney Beans)            0.636364   
4                  ( Onion)                  ( Eggs)            0.272727   
5                   ( Eggs)                 ( Onion)            0.636364   
6                   ( Eggs)                 ( Sugar)            0.636364   
7                  ( Sugar)                  ( Eggs)            0.272727   
8           ( Kidney Beans)                 ( Onion)            0.454545   
9                  ( Onion)          ( Kidney Beans)            0.272727   
10                ( Yogurt)          ( Kidney Beans)            0.272727   
11          ( Kidney Beans)                ( Yogurt)            0.454545   
12          ( Kidney Beans)                   (Milk)            0.454545   
13                   (Milk)          ( Kidney Beans)            0.272727   
14   ( Kidney Beans,  Eggs)                 ( Onion)            0.363636   
15  ( Kidney Beans,  Onion)                  ( Eggs)            0.272727   
16          ( Onion,  Eggs)          ( Kidney Beans)            0.272727   
17          ( Kidney Beans)          ( Onion,  Eggs)            0.454545   
18                  ( Eggs)  ( Kidney Beans,  Onion)            0.636364   
19                 ( Onion)   ( Kidney Beans,  Eggs)            0.272727   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.272727  0.272727    1.000000  3.666667  0.198347         inf  
1             0.272727  0.272727    1.000000  3.666667  0.198347         inf  
2             0.636364  0.363636    0.800000  1.257143  0.074380    1.818182  
3             0.454545  0.363636    0.571429  1.257143  0.074380    1.272727  
4             0.636364  0.272727    1.000000  1.571429  0.099174         inf  
5             0.272727  0.272727    0.428571  1.571429  0.099174    1.272727  
6             0.272727  0.272727    0.428571  1.571429  0.099174    1.272727  
7             0.636364  0.272727    1.000000  1.571429  0.099174         inf  
8             0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
9             0.454545  0.272727    1.000000  2.200000  0.148760         inf  
10            0.454545  0.272727    1.000000  2.200000  0.148760         inf  
11            0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
12            0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
13            0.454545  0.272727    1.000000  2.200000  0.148760         inf  
14            0.272727  0.272727    0.750000  2.750000  0.173554    2.909091  
15            0.636364  0.272727    1.000000  1.571429  0.099174         inf  
16            0.454545  0.272727    1.000000  2.200000  0.148760         inf  
17            0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
18            0.272727  0.272727    0.428571  1.571429  0.099174    1.272727  
19            0.363636  0.272727    1.000000  2.750000  0.173554         inf


The first row (Apple Juice & Baby food) show a LIFT of 3.67 which is a positive correlation between the two items. 

In [ ]:
#Let's view our interpretation values using the Association rule function
df_ar=association_rules(df,metric="conviction", min_threshold=0.8)
df_ar

antecedents              consequents  antecedent support  \
0            ( Apple Juice)             ( Baby Food)            0.272727   
1              ( Baby Food)           ( Apple Juice)            0.272727   
2           ( Kidney Beans)                  ( Eggs)            0.454545   
3                   ( Eggs)          ( Kidney Beans)            0.636364   
4                  ( Onion)                  ( Eggs)            0.272727   
5                   ( Eggs)                 ( Onion)            0.636364   
6                   ( Eggs)                 ( Sugar)            0.636364   
7                  ( Sugar)                  ( Eggs)            0.272727   
8           ( Kidney Beans)                 ( Onion)            0.454545   
9                  ( Onion)          ( Kidney Beans)            0.272727   
10                ( Yogurt)          ( Kidney Beans)            0.272727   
11          ( Kidney Beans)                ( Yogurt)            0.454545   
12          ( Kidney Beans)                   (Milk)            0.454545   
13                   (Milk)          ( Kidney Beans)            0.272727   
14   ( Kidney Beans,  Eggs)                 ( Onion)            0.363636   
15  ( Kidney Beans,  Onion)                  ( Eggs)            0.272727   
16          ( Onion,  Eggs)          ( Kidney Beans)            0.272727   
17          ( Kidney Beans)          ( Onion,  Eggs)            0.454545   
18                  ( Eggs)  ( Kidney Beans,  Onion)            0.636364   
19                 ( Onion)   ( Kidney Beans,  Eggs)            0.272727   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.272727  0.272727    1.000000  3.666667  0.198347         inf  
1             0.272727  0.272727    1.000000  3.666667  0.198347         inf  
2             0.636364  0.363636    0.800000  1.257143  0.074380    1.818182  
3             0.454545  0.363636    0.571429  1.257143  0.074380    1.272727  
4             0.636364  0.272727    1.000000  1.571429  0.099174         inf  
5             0.272727  0.272727    0.428571  1.571429  0.099174    1.272727  
6             0.272727  0.272727    0.428571  1.571429  0.099174    1.272727  
7             0.636364  0.272727    1.000000  1.571429  0.099174         inf  
8             0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
9             0.454545  0.272727    1.000000  2.200000  0.148760         inf  
10            0.454545  0.272727    1.000000  2.200000  0.148760         inf  
11            0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
12            0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
13            0.454545  0.272727    1.000000  2.200000  0.148760         inf  
14            0.272727  0.272727    0.750000  2.750000  0.173554    2.909091  
15            0.636364  0.272727    1.000000  1.571429  0.099174         inf  
16            0.454545  0.272727    1.000000  2.200000  0.148760         inf  
17            0.272727  0.272727    0.600000  2.200000  0.148760    1.818182  
18            0.272727  0.272727    0.428571  1.571429  0.099174    1.272727  
19            0.363636  0.272727    1.000000  2.750000  0.173554         inf

Conviction is similar to LIFT however Conviction will give you a directed measure meaning that on row 3 (Kidney Beans-> Eggs) has a higher conviction than (Eggs -> Kidney Beans) thus it is more likely that the first instance (Kidney Beans -> Eggs) will lead to a purchase. 

In [ ]:
# Second Program

#Loading the dataset
df=pd.read_csv('Dataset2.csv')
df.head()

Shoes, Socks, T-shirt, Pants
0  Sweater, T-Shirt, Hat, Glasses
1   Beanie, T-shirt, Hat, Sweater
2   Sweater, T-Shirt, Hat, Beanie
3  Beanie, Hat, T-Shirt,  Sweater
4    Shoes, Socks, Pants, T-Shirt

In [ ]:
#Since it's missing a column header, so we create new column header name
df=pd.read_csv('Dataset2.csv', names=['Mall'], sep=',')
df.head(10)

Mall
0    Shoes, Socks, T-shirt, Pants
1  Sweater, T-Shirt, Hat, Glasses
2   Beanie, T-shirt, Hat, Sweater
3   Sweater, T-Shirt, Hat, Beanie
4  Beanie, Hat, T-Shirt,  Sweater
5    Shoes, Socks, Pants, T-Shirt

In [ ]:
#data size
df.shape

(6, 1)

In [ ]:
#we are turning the pandas dataframe into a list of lists
data=list(df["Mall"].apply(lambda x:x.split(",")))
data

[['Shoes', ' Socks', ' T-shirt', ' Pants'],
 ['Sweater', ' T-Shirt', ' Hat', ' Glasses'],
 ['Beanie', ' T-shirt', ' Hat', ' Sweater'],
 ['Sweater', ' T-Shirt', ' Hat', ' Beanie'],
 ['Beanie', ' Hat', ' T-Shirt', '  Sweater'],
 ['Shoes', ' Socks', ' Pants', ' T-Shirt']]

In [ ]:
#Let's transform the list with one-hot encoding
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)
df=df.replace(True,1)
df=df.replace(False,0)
df

Sweater   Beanie   Glasses   Hat   Pants   Socks   Sweater   T-Shirt  \
0          0        0         0     0       1       1         0         0   
1          0        0         1     1       0       0         0         1   
2          0        0         0     1       0       0         1         0   
3          0        1         0     1       0       0         0         1   
4          1        0         0     1       0       0         0         1   
5          0        0         0     0       1       1         0         1   

    T-shirt  Beanie  Shoes  Sweater  
0         1       0      1        0  
1         0       0      0        1  
2         1       1      0        0  
3         0       0      0        1  
4         0       1      0        0  
5         0       0      1        0

In [ ]:
#Set a threshold value for the support value and calculate the support value
df=apriori(df,min_support=0.2, use_colnames=True)
df

support                   itemsets
0   0.666667                     ( Hat)
1   0.333333                   ( Pants)
2   0.333333                   ( Socks)
3   0.666667                 ( T-Shirt)
4   0.333333                 ( T-shirt)
5   0.333333                   (Beanie)
6   0.333333                    (Shoes)
7   0.333333                  (Sweater)
8   0.500000           ( Hat,  T-Shirt)
9   0.333333             ( Hat, Beanie)
10  0.333333            ( Hat, Sweater)
11  0.333333           ( Socks,  Pants)
12  0.333333            ( Pants, Shoes)
13  0.333333            ( Socks, Shoes)
14  0.333333        ( T-Shirt, Sweater)
15  0.333333  ( Hat,  T-Shirt, Sweater)
16  0.333333    ( Socks,  Pants, Shoes)

Any item or combinations less than 2.0 are not shown.

In [ ]:
#Here is the CONFIDENCE (Try different min_threshold value to explore this more)
df_ar=association_rules(df,metric="confidence", min_threshold=0.3)
df_ar

antecedents          consequents  antecedent support  \
0                ( Hat)           ( T-Shirt)            0.666667   
1            ( T-Shirt)               ( Hat)            0.666667   
2                ( Hat)             (Beanie)            0.666667   
3              (Beanie)               ( Hat)            0.333333   
4                ( Hat)            (Sweater)            0.666667   
5             (Sweater)               ( Hat)            0.333333   
6              ( Socks)             ( Pants)            0.333333   
7              ( Pants)             ( Socks)            0.333333   
8              ( Pants)              (Shoes)            0.333333   
9               (Shoes)             ( Pants)            0.333333   
10             ( Socks)              (Shoes)            0.333333   
11              (Shoes)             ( Socks)            0.333333   
12           ( T-Shirt)            (Sweater)            0.666667   
13            (Sweater)           ( T-Shirt)            0.333333   
14     ( Hat,  T-Shirt)            (Sweater)            0.500000   
15      ( Hat, Sweater)           ( T-Shirt)            0.333333   
16  ( T-Shirt, Sweater)               ( Hat)            0.333333   
17               ( Hat)  ( T-Shirt, Sweater)            0.666667   
18           ( T-Shirt)      ( Hat, Sweater)            0.666667   
19            (Sweater)     ( Hat,  T-Shirt)            0.333333   
20     ( Socks,  Pants)              (Shoes)            0.333333   
21      ( Socks, Shoes)             ( Pants)            0.333333   
22      ( Pants, Shoes)             ( Socks)            0.333333   
23             ( Socks)      ( Pants, Shoes)            0.333333   
24             ( Pants)      ( Socks, Shoes)            0.333333   
25              (Shoes)     ( Socks,  Pants)            0.333333   

    consequent support   support  confidence   lift  leverage  conviction  
0             0.666667  0.500000    0.750000  1.125  0.055556    1.333333  
1             0.666667  0.500000    0.750000  1.125  0.055556    1.333333  
2             0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
3             0.666667  0.333333    1.000000  1.500  0.111111         inf  
4             0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
5             0.666667  0.333333    1.000000  1.500  0.111111         inf  
6             0.333333  0.333333    1.000000  3.000  0.222222         inf  
7             0.333333  0.333333    1.000000  3.000  0.222222         inf  
8             0.333333  0.333333    1.000000  3.000  0.222222         inf  
9             0.333333  0.333333    1.000000  3.000  0.222222         inf  
10            0.333333  0.333333    1.000000  3.000  0.222222         inf  
11            0.333333  0.333333    1.000000  3.000  0.222222         inf  
12            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
13            0.666667  0.333333    1.000000  1.500  0.111111         inf  
14            0.333333  0.333333    0.666667  2.000  0.166667    2.000000  
15            0.666667  0.333333    1.000000  1.500  0.111111         inf  
16            0.666667  0.333333    1.000000  1.500  0.111111         inf  
17            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
18            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
19            0.500000  0.333333    1.000000  2.000  0.166667         inf  
20            0.333333  0.333333    1.000000  3.000  0.222222         inf  
21            0.333333  0.333333    1.000000  3.000  0.222222         inf  
22            0.333333  0.333333    1.000000  3.000  0.222222         inf  
23            0.333333  0.333333    1.000000  3.000  0.222222         inf  
24            0.333333  0.333333    1.000000  3.000  0.222222         inf  
25            0.333333  0.333333    1.000000  3.000  0.222222         inf

The first row shows (Hat -> T-Shirt) shows a confidence of .75 which means that these two items are bought together 75% of the time. The third row (Hat -> Beanie) shows a condidence level of .50 which means that these two items are bought together 50% of the time. 

In [ ]:
#Here is LIFT (try different min_threshold value to explore more) 
df_ar=association_rules(df,metric="lift", min_threshold=0.8)
df_ar

antecedents          consequents  antecedent support  \
0                ( Hat)           ( T-Shirt)            0.666667   
1            ( T-Shirt)               ( Hat)            0.666667   
2                ( Hat)             (Beanie)            0.666667   
3              (Beanie)               ( Hat)            0.333333   
4                ( Hat)            (Sweater)            0.666667   
5             (Sweater)               ( Hat)            0.333333   
6              ( Socks)             ( Pants)            0.333333   
7              ( Pants)             ( Socks)            0.333333   
8              ( Pants)              (Shoes)            0.333333   
9               (Shoes)             ( Pants)            0.333333   
10             ( Socks)              (Shoes)            0.333333   
11              (Shoes)             ( Socks)            0.333333   
12           ( T-Shirt)            (Sweater)            0.666667   
13            (Sweater)           ( T-Shirt)            0.333333   
14     ( Hat,  T-Shirt)            (Sweater)            0.500000   
15      ( Hat, Sweater)           ( T-Shirt)            0.333333   
16  ( T-Shirt, Sweater)               ( Hat)            0.333333   
17               ( Hat)  ( T-Shirt, Sweater)            0.666667   
18           ( T-Shirt)      ( Hat, Sweater)            0.666667   
19            (Sweater)     ( Hat,  T-Shirt)            0.333333   
20     ( Socks,  Pants)              (Shoes)            0.333333   
21      ( Socks, Shoes)             ( Pants)            0.333333   
22      ( Pants, Shoes)             ( Socks)            0.333333   
23             ( Socks)      ( Pants, Shoes)            0.333333   
24             ( Pants)      ( Socks, Shoes)            0.333333   
25              (Shoes)     ( Socks,  Pants)            0.333333   

    consequent support   support  confidence   lift  leverage  conviction  
0             0.666667  0.500000    0.750000  1.125  0.055556    1.333333  
1             0.666667  0.500000    0.750000  1.125  0.055556    1.333333  
2             0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
3             0.666667  0.333333    1.000000  1.500  0.111111         inf  
4             0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
5             0.666667  0.333333    1.000000  1.500  0.111111         inf  
6             0.333333  0.333333    1.000000  3.000  0.222222         inf  
7             0.333333  0.333333    1.000000  3.000  0.222222         inf  
8             0.333333  0.333333    1.000000  3.000  0.222222         inf  
9             0.333333  0.333333    1.000000  3.000  0.222222         inf  
10            0.333333  0.333333    1.000000  3.000  0.222222         inf  
11            0.333333  0.333333    1.000000  3.000  0.222222         inf  
12            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
13            0.666667  0.333333    1.000000  1.500  0.111111         inf  
14            0.333333  0.333333    0.666667  2.000  0.166667    2.000000  
15            0.666667  0.333333    1.000000  1.500  0.111111         inf  
16            0.666667  0.333333    1.000000  1.500  0.111111         inf  
17            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
18            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
19            0.500000  0.333333    1.000000  2.000  0.166667         inf  
20            0.333333  0.333333    1.000000  3.000  0.222222         inf  
21            0.333333  0.333333    1.000000  3.000  0.222222         inf  
22            0.333333  0.333333    1.000000  3.000  0.222222         inf  
23            0.333333  0.333333    1.000000  3.000  0.222222         inf  
24            0.333333  0.333333    1.000000  3.000  0.222222         inf  
25            0.333333  0.333333    1.000000  3.000  0.222222         inf

The first row (Hat & T-Shirt) shows a LIFT of 1.125 which is a slight positive correlation between the two items. Row 7 (Socks & Pants) shows a significant positive correlation of 3 between the two items. 

In [ ]:
#Let's view our interpretation values using the Association rule function
df_ar=association_rules(df,metric="conviction", min_threshold=0.8)
df_ar

antecedents          consequents  antecedent support  \
0                ( Hat)           ( T-Shirt)            0.666667   
1            ( T-Shirt)               ( Hat)            0.666667   
2                ( Hat)             (Beanie)            0.666667   
3              (Beanie)               ( Hat)            0.333333   
4                ( Hat)            (Sweater)            0.666667   
5             (Sweater)               ( Hat)            0.333333   
6              ( Socks)             ( Pants)            0.333333   
7              ( Pants)             ( Socks)            0.333333   
8              ( Pants)              (Shoes)            0.333333   
9               (Shoes)             ( Pants)            0.333333   
10             ( Socks)              (Shoes)            0.333333   
11              (Shoes)             ( Socks)            0.333333   
12           ( T-Shirt)            (Sweater)            0.666667   
13            (Sweater)           ( T-Shirt)            0.333333   
14     ( Hat,  T-Shirt)            (Sweater)            0.500000   
15      ( Hat, Sweater)           ( T-Shirt)            0.333333   
16  ( T-Shirt, Sweater)               ( Hat)            0.333333   
17               ( Hat)  ( T-Shirt, Sweater)            0.666667   
18           ( T-Shirt)      ( Hat, Sweater)            0.666667   
19            (Sweater)     ( Hat,  T-Shirt)            0.333333   
20     ( Socks,  Pants)              (Shoes)            0.333333   
21      ( Socks, Shoes)             ( Pants)            0.333333   
22      ( Pants, Shoes)             ( Socks)            0.333333   
23             ( Socks)      ( Pants, Shoes)            0.333333   
24             ( Pants)      ( Socks, Shoes)            0.333333   
25              (Shoes)     ( Socks,  Pants)            0.333333   

    consequent support   support  confidence   lift  leverage  conviction  
0             0.666667  0.500000    0.750000  1.125  0.055556    1.333333  
1             0.666667  0.500000    0.750000  1.125  0.055556    1.333333  
2             0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
3             0.666667  0.333333    1.000000  1.500  0.111111         inf  
4             0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
5             0.666667  0.333333    1.000000  1.500  0.111111         inf  
6             0.333333  0.333333    1.000000  3.000  0.222222         inf  
7             0.333333  0.333333    1.000000  3.000  0.222222         inf  
8             0.333333  0.333333    1.000000  3.000  0.222222         inf  
9             0.333333  0.333333    1.000000  3.000  0.222222         inf  
10            0.333333  0.333333    1.000000  3.000  0.222222         inf  
11            0.333333  0.333333    1.000000  3.000  0.222222         inf  
12            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
13            0.666667  0.333333    1.000000  1.500  0.111111         inf  
14            0.333333  0.333333    0.666667  2.000  0.166667    2.000000  
15            0.666667  0.333333    1.000000  1.500  0.111111         inf  
16            0.666667  0.333333    1.000000  1.500  0.111111         inf  
17            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
18            0.333333  0.333333    0.500000  1.500  0.111111    1.333333  
19            0.500000  0.333333    1.000000  2.000  0.166667         inf  
20            0.333333  0.333333    1.000000  3.000  0.222222         inf  
21            0.333333  0.333333    1.000000  3.000  0.222222         inf  
22            0.333333  0.333333    1.000000  3.000  0.222222         inf  
23            0.333333  0.333333    1.000000  3.000  0.222222         inf  
24            0.333333  0.333333    1.000000  3.000  0.222222         inf  
25            0.333333  0.333333    1.000000  3.000  0.222222         inf

Conviction is similar to LIFT however Conviction will give you a directed measure meaning that on row 3 (Hat -> Beanie) has a higher conviction than (Beanie -> Hat) thus it is more likely that the first instance (Hat -> Beanie) will lead to a purchase. 